File: Data Structures/src/medium/Min difference queries.java
import java.io.*;
import java.util.*;

public class Main {

    static final int MAX = 100002;
    static final int MAGIC = 333;
    static final int MM = 500;

    static int n, q;

    static int[] v = new int[MAX];
    static int[][] u = new int[MAGIC][MM];
    static int[] sz = new int[MAGIC];

    static boolean[] ex = new boolean[MAX];

    static ArrayList<Integer> vv = new ArrayList<>();
    static ArrayList<Integer>[] idx = new ArrayList[MAX];

    static class FastScanner {
        private final byte[] buffer = new byte[1 << 16];
        private int ptr = 0, len = 0;
        private final InputStream in;

        FastScanner(InputStream in) {
            this.in = in;
        }

        int read() throws IOException {
            if (ptr >= len) {
                len = in.read(buffer);
                ptr = 0;
                if (len <= 0) return -1;
            }
            return buffer[ptr++];
        }

        int nextInt() throws IOException {
            int c;
            while ((c = read()) <= ' ') ;
            int sign = 1;
            if (c == '-') {
                sign = -1;
                c = read();
            }
            int val = 0;
            while (c > ' ') {
                val = val * 10 + (c - '0');
                c = read();
            }
            return val * sign;
        }
    }

    static int lowerBound(ArrayList<Integer> arr, int x) {
        int l = 0, r = arr.size();
        while (l < r) {
            int m = (l + r) >>> 1;
            if (arr.get(m) < x) l = m + 1;
            else r = m;
        }
        return l;
    }

    static int upperBound(ArrayList<Integer> arr, int x) {
        int l = 0, r = arr.size();
        while (l < r) {
            int m = (l + r) >>> 1;
            if (arr.get(m) <= x) l = m + 1;
            else r = m;
        }
        return l;
    }

    public static void main(String[] args) throws Exception {
        FastScanner fs = new FastScanner(System.in);

        n = fs.nextInt();
        q = fs.nextInt();

        for (int i = 0; i < MAX; i++) idx[i] = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            int a = fs.nextInt();
            v[i] = a;
            vv.add(a);
        }

        // Coordinate compression
        Collections.sort(vv);
        for (int i = 0; i < n; i++) {
            v[i] = Collections.binarySearch(vv, v[i]);
            idx[v[i]].add(i);
        }

        // Build sqrt blocks
        for (int i = 0; i < n; i += MAGIC) {
            Arrays.fill(ex, false);
            int block = i / MAGIC;
            for (int j = i; j < n; j++) {
                if (ex[v[j]]) continue;
                ex[v[j]] = true;
                u[block][sz[block]++] = v[j];
                if (sz[block] == MM) break;
            }
        }

        int last = 0;
        StringBuilder out = new StringBuilder();

        while (q-- > 0) {
            int a = fs.nextInt();
            int b = fs.nextInt();

            int l = (a + last) % n;
            int r = (b + last) % n;

            if (l > r) {
                int t = l;
                l = r;
                r = t;
            }

            int belong = l / MAGIC;
            TreeSet<Integer> set = new TreeSet<>();
            boolean bad = false;

            for (int i = 0; i < sz[belong]; i++) {
                int val = u[belong][i];
                ArrayList<Integer> list = idx[val];

                int lef = lowerBound(list, l);
                int rig = upperBound(list, r);
                int oc = rig - lef;

                if (oc == 0) continue;
                if (set.contains(oc)) {
                    bad = true;
                    break;
                }
                set.add(oc);
            }

            if (bad) {
                out.append("0\n");
                last = 0;
                continue;
            }

            if (set.size() == 1) {
                out.append("-1\n");
                last = -1;
                continue;
            }

            int ans = Integer.MAX_VALUE;
            int prev = -1;
            for (int x : set) {
                if (prev != -1) ans = Math.min(ans, x - prev);
                prev = x;
            }

            out.append(ans).append('\n');
            last = ans;
        }

        System.out.print(out);
    }
}

File: Data Structures/src/medium/Min difference queries.cpp
#include "bits/stdc++.h"
using namespace std;
 
#define MAX 100002
 
int n;
int q;
 
vector<int> v;
 
#define MAGIC 333
#define MM 500
int u[MAGIC][MM];
 
int sz[MAGIC];
 
 
bool ex[MAX];
 
vector<int> vv;
 
vector<int> idx[MAX];
 
set<int> s;
 
 
int main() {
	cin >> n >> q;
	assert(1<=n&&n<=100000);
	assert(1<=q&&q<=100000);
	for (int i = 0; i < n; i++) {
		int a;
		scanf("%d", &a);
		assert(1<=a);
		assert(a<=1000000000);
		v.push_back(a);
		vv.push_back(a);
	}
	sort(vv.begin(), vv.end());
	for (int i = 0; i < v.size(); i++) {
		v[i] = lower_bound(vv.begin(), vv.end(), v[i]) - vv.begin();
		idx[v[i]].push_back(i);
	}
	for (int i = 0; i < n; i+=MAGIC) {
		memset(ex, false, sizeof(ex));
		for (int j = i; j < n; j++) {
			if (ex[v[j]])continue;
			ex[v[j]] = true;
			u[i / MAGIC][sz[i / MAGIC]] = v[j];
			sz[i / MAGIC]++;
			if(sz[i/MAGIC]==MM)break;
		}
	}
	int las = 0;
	while (q--) {
		int a,b;
		scanf("%d%d", &a, &b);
		assert(1<=a&&a<=n&&1<=b&&b<=n);
		int l = (a + las) % n + 1;
		int r = (b + las) % n + 1;
		l--;
		r--;
		assert(0<=l&&0<=r);
		int belong = l / MAGIC;
		s.clear();
		bool en = false;
		for (int i = 0; i < sz[belong]; i++) {  
			int val = u[belong][i];
			int lef = lower_bound(idx[val].begin(), idx[val].end(), l) - idx[val].begin();
			int rig = upper_bound(idx[val].begin(), idx[val].end(), r) - idx[val].begin();
			int oc = rig - lef;
			if (oc == 0)continue;
			if (s.count(oc)) {
				en = true;
				break;
			}
			s.insert(oc);
		}
		if (en) {
			puts("0");
			las = 0;
			continue;
		}
		if (s.size() == 1) {
			puts("-1");
			las = -1;
			continue;
		}
		int ans = INT_MAX;
		int pr = -1;
		for (auto el : s) {
			if (pr != -1) {
				ans = min(ans, el - pr);
			}
			pr = el;
		}
		printf("%d\n", ans);
		las = ans;
	}
	return 0;
}
